### PR TITLE
[TTAHUB-2575] Escape fields for XSS

### PR DESCRIPTION
**Pull request by thewatermethod**
## Description of change
Escape the fields called out via pen test

This PR is large because I had to relocate the hooks folder to mitigate OOM issues with Jest on the CI. The vast majority of files were only slightly changed to update imports. Also, the relocation caused test coverage to fall below acceptable levels, so I had to add ~a few~ many tests as well.

## How to test

- Follow the instructions in the pen test tickets
- I self-reviewed the files that were changed as part of the Jira. The rest of the files, nearly 140, are just renames or added tests.

## Issue(s)

* https://jira.acf.gov/browse/TTAHUB-2573
* https://jira.acf.gov/browse/TTAHUB-2575
* https://jira.acf.gov/browse/TTAHUB-2576
* https://jira.acf.gov/browse/TTAHUB-2577

### Every PR

- [ ] Meets issue criteria
- [ ] JIRA ticket status updated
- [ ] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions
- [ ] UI review complete

### Before merge to main

- [ ] OHS demo complete
- [ ] Ready to create production PR

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -52,7 +52,7 @@
     "db:seed:prod": "node_modules/.bin/sequelize db:seed:all --options-path .production.sequelizerc",
     "db:seed:undo": "node_modules/.bin/sequelize db:seed:undo:all",
     "db:seed:undo:prod": "node_modules/.bin/sequelize db:seed:undo:all --options-path .production.sequelizerc",
-    "docker:deps": "docker-compose run --rm backend yarn install && docker-compose run --rm frontend yarn install",
+    "docker:deps": "docker-compose run --rm backend yarn install && docker-compose run --rm frontend yarn install && docker-compose run --rm testingonly yarn install",
```
Comment (thewatermethod) [inline package.json:55]: noticed this was missing

File: bin/test-backend-ci [hunk 2]
```diff
@@ -41,7 +41,7 @@ main(){
 	check_exit "$?"
 
     # then list through the folders and run the tests
-    targets=("lib" "middleware" "models" "policies" "routes" "scopes" "services" "tools" "widgets" "goalServices")
+    targets=("lib" "middleware" "models" "policies" "routes" "scopes" "services" "tools" "widgets" "goalServices" "hooks")
```
Comment (thewatermethod) [inline bin/test-backend-ci:44]: test models and hooks separately on CI

File: packages/common/README.md [hunk 3]
```diff
@@ -21,6 +21,9 @@ Note: On Windows you will need to use `yarn add @ttahub/common@1.x.0` to update
 
 ## Versions
 
+## 2.0.19
```
Comment (thewatermethod) [inline packages/common/README.md:24]: I wrote an escapeHTML function and then removed it in favor of using lodash's.

Comment (thewatermethod) [reply packages/common/README.md:24]: Hence the version number jumps

File: src/models/helpers/safeParse.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+function safeParse(instance) {
```
Comment (thewatermethod) [inline src/models/helpers/safeParse.js:1]: This function already existed, I just moved it to it's own file

File: src/models/helpers/escapeFields.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
+import { escape } from 'lodash';
+import { auditLogger } from '../../logger';
+import safeParse from './safeParse';
+
```
Comment (thewatermethod) [inline src/models/helpers/escapeFields.js:4]: The actual new code from this PR: a function to escape a given field, or a given field within a data JSONB object

File: src/hooks/activityReport.js [hunk 6]
```diff
@@ -1,31 +1,35 @@
...
+const { formatModelForAwsElasticsearch } = require('../lib/awsElasticSearch/modelMapper');
+const { addIndexDocument, deleteIndexDocument } = require('../lib/awsElasticSearch/index');
 const {
   findOrCreateCollaborator,
   removeCollaboratorsForType,
-} = require('../helpers/genericCollaborator');
+} = require('../models/helpers/genericCollaborator');
 const { destroyLinkedSimilarityGroups } = require('./activityReportGoal');
 
+const AR_FIELDS_TO_ESCAPE = ['additionalNotes', 'context'];
```
Comment (thewatermethod) [inline src/hooks/activityReport.js:28]: any fields added here will be escaped

File: src/hooks/activityReportApprover.js [hunk 7]
```diff
@@ -1,4 +1,7 @@
 const { APPROVER_STATUSES, REPORT_STATUSES } = require('@ttahub/common');
+const { default: escapeFields } = require('../models/helpers/escapeFields');
+
+const FIELDS_TO_ESCAPE = ['note'];
```
Comment (thewatermethod) [inline src/hooks/activityReportApprover.js:4]: any fields added here will be escaped

File: src/hooks/eventReportPilot.js [hunk 8]
```diff
@@ -1,28 +1,16 @@
...
-  if (instance?.data) {
-    return instance.data;
-  }
+const { auditLogger } = require('../logger');
+const { escapeDataFields } = require('../models/helpers/escapeFields');
+const safeParse = require('../models/helpers/safeParse');
 
-  return null;
-};
+const fieldsToEscape = ['eventName'];
```
Comment (thewatermethod) [inline src/hooks/eventReportPilot.js:13]: any fields added here will be escaped

**Comment by GarrettEHill** [comment]
Could we try changing the resource class of the backend test to "resource_class: large"

**Comment by thewatermethod** [comment]
> Could we try changing the resource class of the backend test to "resource_class: large"

we can try, not sure I want to go through moving it all back tho

**Comment by GarrettEHill** [comment]
> > Could we try changing the resource class of the backend test to "resource_class: large"
> 
> we can try, not sure I want to go through moving it all back tho
> 
> **edit** nvm, @GarrettEHill, it already is

sorry xlarge

**Comment by thewatermethod** [comment]
yeah sure

**Comment by thewatermethod** [comment]
in case anyone is interested, [it looks like we'd have to pay to get xlarge](https://circleci.com/product/features/resource-classes/)

**Review by nvms — APPROVED** [review]
LGTM!

**Pull request merged**